src/github.com/matrix-org/dendrite/roomserver/storage/sql.go

```diff
@@ -494,3 +555,182 @@ func (s *statements) bulkSelectEventJSON(eventNIDs []int64) ([]eventJSONPair, er
 	}
 	return results[:i], nil
 }
+
+const stateSchema = `
+-- The state of a room before an event.
+-- Stored as a list of state_data entries stored in a separate table.
+-- The actual state is constructed by combining all the state_data entries
+-- referenced by state_data_nids together. If the same state key tuple appears
+-- multiple times then the entry from the later state_data clobbers the earlier
+-- entries.
+-- This encoding format allows us to implement a delta encoding which is useful
+-- because room state tends to accumulate small changes over time. Although if
+-- the list of deltas becomes too long it becomes more efficient to encode
+-- the full state under single state_data_nid.
+CREATE SEQUENCE IF NOT EXISTS state_nid_seq;
+CREATE TABLE IF NOT EXISTS state (
+    -- Local numeric ID for the state.
+    state_nid bigint PRIMARY KEY DEFAULT nextval('state_nid_seq'),
+    -- Local numeric ID of the room this state is for.
+    -- Unused in normal operation, but useful for background work or ad-hoc debugging.
+    room_nid bigint NOT NULL,
+    -- List of state_data_nids, stored sorted by state_data_nid.
+    state_data_nids bigint[] NOT NULL
+);
+`
+
+const insertStateSQL = "" +
+	"INSERT INTO state (room_nid, state_data_nids)" +
+	" VALUES ($1, $2)" +
+	" RETURNING state_nid"
+
+const bulkSelectStateDataNIDsSQL = "" +
+	"SELECT state_nid, state_data_nids FROM state" +
+	" WHERE state_nid = ANY($1) ORDER BY state_nid"
```

Comment:
This ORDER BY is going to be used for clobbering right? Maybe? Shame I don't know the significance of it because there is no comment.